ts/calculator.ts

```diff
@@ -0,0 +1,476 @@
+        }
+
+        ans = result;
+        data.operation = [result];
+        data.formula = [result];
+        updateOutputresult(result);
+        return;
+    }
+    updateOutputOperation(data.operation.join(""));
+}
+interface ReplaceType {
+    toReplace: string,
+    replacement: string
+}
+function factorialnumgetter(formula: (string | number | boolean)[], FACTORIAL_SEARCH_RESULT: number[]) {
+    let numbers: ReplaceType[] = [];
+
+    FACTORIAL_SEARCH_RESULT.forEach((fact_index) => {
+        let number = [];
+        let prev_idx = fact_index - 1;
+
+        while (prev_idx >= 0) {
+            let is_operator = false;
+
+            OPERATORS.forEach((OPERATOR) => {
+                if (formula[prev_idx] === OPERATOR) {
+                    is_operator = true;
+                }
+            });
+
+            if ((is_operator)) {
+                break;
+            }
+
+            number.unshift(formula[prev_idx]);
+            prev_idx--;
+        }
+
+        let number_str = number.join("");
+        const factorial = "factorial(",
+            close_paren = ")";
+        let toreplace = number_str + FACTORIAL;
+        let replacement = factorial + number_str + close_paren;
+        numbers.push({
+            toReplace: toreplace,
+            replacement: replacement,
+        });
+    });
+    return numbers;
+}
+
+function powerbasegetter(formula: (string | number | boolean)[], POWER_SEARCH_RESULT: number[]) {
+    let powers_base: (number | string)[] = [];
+
+    POWER_SEARCH_RESULT.forEach((power_index) => {
+        let base = [];
+        let paren_count = 0;
+        let prev_idx = power_index - 1;
+
+        while (prev_idx >= 0) {
+            let is_operator = false;
+
+            if (formula[prev_idx] === "(") {
+                paren_count -= 1;
+            }
+
+            if (formula[prev_idx] === ")") {
+                paren_count += 1;
+            }
+
+            OPERATORS.forEach((OPERATOR) => {
+                if (formula[prev_idx] === OPERATOR) {
+                    is_operator = true;
+                }
+            });
+
+            let is_power = formula[prev_idx] === POWER;
+
+            if ((is_operator && paren_count === 0) || is_power) {
+                break;
+            }
+
+            base.unshift(formula[prev_idx]);
+            prev_idx--;
+        }
+        powers_base.push(base.join(""));
+    });
+
+    return powers_base;
+}
+
+function search(array: (string | number | boolean)[], keyword: string) {
+    let search_res: number[] = [];
+
+    array.forEach((element, index) => {
+        if (element == keyword) {
+            search_res.push(index);
+        }
+    });
+
+    return search_res;
+}
+
+function updateOutputOperation(operation: string) {
+    output_operation_element.value = operation;
+}
+
+function updateOutputresult(result: (number | string)) {
+    output_operation_element.value = result.toString();
+}
+
+function trigo(callback: Function, angle: number) {
```

Comment:
Instead of `Function`, Add type definition to specify arguments and return type of callback function.